### PR TITLE
Full Glm support for SAG

### PR DESCRIPTION
**Pull request by IshmaelBelghazi**
@tdhock, I have replicated part of Mark's matlab examples in demo/example_SAG2.R. There is also a demo to monitoring the behavior of training cost and gradient norm during training. So far the results seem to be consistent. I'll add the graphs asap. 

Also the stopping condition seems to be tricky. To reduce storage cost, mark's only stores the gradient with respect to the inner product of the weight and the example (O(n) vs O(np)). So just checking the norm of the corresponding variable (g in the code) won't be enough. I am thining of adding another variable that will be a running sum of the approximate gradient with respect to the weights. 

Finally, even if the results seem to be consitent and agree perfectly with Mark's example files, there seems to not be consistent with glmnet. Actually, they are way off. I think I might be making a mistake with glmnet somewhere in the tests.

### CODE REVIEW

**Comment by tdhock** [comment]
maybe it has something to do with the fact that glmnet normalizes the
features internally (subtract the mean and divide by the standard deviation
for each feature I think). Do you normalize first? you may want to read the
glmnet documentation to see if you are checking normalized weights against
un-normalized weights.

**Comment by IshmaelBelghazi** [comment]
Yes, I normalize the features beforehand and pass FALSE to standardize in glmnet.

**Comment by tdhock** [comment]
Great to see that you have started computing the regularized model based on the glmnet regularization parameter choices, in demos/monitoring_training.R

However when I ran that code I got 

```
[1] "Processing training period: 1/100"
[1] "Fitting SAG with constant step size"
[1] "Fitting SAG with linesearch"
Error: could not find function ".bernoulli_loss"
```

File: R/glm_models.R [hunk 1]
```diff
@@ -1,19 +1,33 @@
...
+## *** Bernoulli
+##' @export
+.bernoulli_loss <- function(X, y, w, lambda=0) {
+  innerProd <- X  %*% w
+  losses <- log(1 + exp(-y * innerProd))
+  loss <- sum(losses)/NROW(X) + 0.5 * lambda * sum(w^2)
+  
+}
+##' @export
+.bernoulli_grad <- function(X, y, w, lambda=0) {
```
Comment (tdhock) [inline R/glm_models.R]: Can you please add the loss and gradient equations in mathematical notation in comments here? And can you please add a reference to the equation number where the loss is defined in the SAG paper?

Comment (IshmaelBelghazi) [inline R/glm_models.R]: Sure. I'll add this asap.

**Comment by IshmaelBelghazi** [comment]
Strange. It works for me. Did you do a `load_all()` before running the
script?

**Comment by IshmaelBelghazi** [comment]
Btw, the code runs slow as I had to use a for loop to compute the gradient
to stay within the available memory limits. I am about to write a C version
of the gradient of the cost right now. This should speed up things
considerable.

**Comment by tdhock** [comment]
no I did not do load_all(), I did library(bigoptim).

**Comment by IshmaelBelghazi** [comment]
That might be it. Maybe you have the master branch installed.
devtools::load_all() should do the job.

**Comment by tdhock** [comment]
I have the glm branch installed.

in practice when people like me use your package they will do library(bigoptim), not load_all(). So if you want users to be able to use your demos your will need to export functions you use in the demos, such as ".bernoulli_loss"

**Comment by IshmaelBelghazi** [comment]
I have updated NAMESPACE. Should work now.
